From 259f6ccf75a628a5c63a6e80427a0c109e6844d1 Mon Sep 17 00:00:00 2001
From: Dimitri John Ledkov <dimitri.ledkov@surgut.co.uk>
Date: Thu, 23 Oct 2025 11:10:06 +0100
Subject: [PATCH 1/2] tlsutil: choose ECDSA in FIPS mode, until 3.4 module

Chainguard FIPS 3.4 will have Ed25519 support, but it is not yet
validated.

Use ECDSA code path until then.
---
 lib/tlsutil/tlsutil.go | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/lib/tlsutil/tlsutil.go b/lib/tlsutil/tlsutil.go
index cbdd05b3b..d9a54c452 100644
--- a/lib/tlsutil/tlsutil.go
+++ b/lib/tlsutil/tlsutil.go
@@ -10,6 +10,7 @@ import (
 	"crypto/ecdsa"
 	"crypto/ed25519"
 	"crypto/elliptic"
+	"crypto/fips140"
 	"crypto/rsa"
 	"crypto/tls"
 	"crypto/x509"
@@ -98,7 +99,10 @@ func generateCertificate(commonName string, lifetimeDays int, compatible bool) (
 	var pub, priv any
 	var err error
 	var sigAlgo x509.SignatureAlgorithm
-	if compatible {
+	// Chainguard FIPS 3.1 module doesn't yet have Ed25519
+	// Chainguard FIPS 3.4 module does
+	// Choose ECDSA-P256 until Chainguard FIPS 3.4 is validated
+	if compatible || fips140.Enabled() {
 		// For browser connections we prefer ECDSA-P256
 		sigAlgo = x509.ECDSAWithSHA256
 		var pk *ecdsa.PrivateKey
-- 
2.51.0

